[addons][filesystem][base] add some new functions, improvement, cleanups and doc rework

AlwinEsch opened the pull request:
## Description
This request was primary thought to update the filesystem parts, but as mostly all addons use them, need also all a update.

For this reason is the request a bit extended to add also other parts, where also related to all addons and to reduce the extreme big update rounds on all 100 addons.
This additional added parts are mostly minor things where not breaks the operation (related to working ways on addon itself).

Also some things why it looks bigger are clang cleanups on the headers and separation to extra "C" files.

Some parts added related to inpustream addons, then fixes about https://github.com/xbmc/xbmc/issues/17514, make the doxygen a bit cleaner.

> **NOTE:** If nothing special comes, should after them the addon main version (`ADDON_GLOBAL_VERSION_MAIN_MIN`) hopefully a bit longer without changes.
Thats why some parts combined in this request, to make all to them on one time.

Docs visible here:
- http://alwinesch.github.io/group__cpp__kodi__vfs.html

Commit 1:
------------------------------
**[addons][filesystem] add function IsOpen to CFile**

This thought to allow other place a check that related file
was really opened.
 
Commit 2:
------------------------------
**[addons][filesystem] replace __stat64 with STAT_STRUCTURE**

This done to have it more OS independent, before was about time in
not complete compatible between.
 
Commit 3:
------------------------------
**[addons][filesystem] make interface "C" conform**

Before was it not in correct "C" ABI between Kodi and Addon.
This add #ifdef about and change some parts on "C" side to allow
compile there.
 
Commit 4:
------------------------------
**[addons][filesystem] cleanup namespace kodi::addon use**

before was on every place this where makes the lines very long,
this change it to "using namespace kodi::addon" everywhere, where
reduce the size.
 
Commit 5:
------------------------------
**[addons][filesystem] clang code cleanup**

this cleanup the parts to match his style
 
Commit 6:
------------------------------
**[addons][filesystem] cleanup documentation**

This makes it cleaner and easier on review. Further are the parts
placed in groups.
 
Commit 7:
------------------------------
**[addons][filesystem] add IoControl functions**

Following are added:
- bool IoControlGetSeekPossible()
- bool IoControlGetCacheStatus(CacheStatus& status)
- bool IoControlSetCacheRate(unsigned int rate)
- bool IoControlSetRetry(bool retry)
 
Commit 8:
------------------------------
**[addons][filesystem] add some functions to check file path relate to web**

Following functions are added:
- bool kodi::vfs::IsInternetStream(const std::string& path, bool strictCheck = false);
- bool kodi::vfs::IsOnLAN(const std::string& path);
- bool kodi::vfs::IsRemote(const std::string& path);
- bool kodi::vfs::IsLocal(const std::string& path);
- bool kodi::vfs::IsURL(const std::string& path);
 
Commit 9:
------------------------------
**[addons][filesystem] separate "C" and "C++" parts of headers**

By them becomes the "C" part placed in own header to see better what language
is where and to make easier for other languages.
 
Commit 10:
------------------------------
**[addons][filesystem] add more functions to get web content**

Following are added:
- GetHttpHeader
  - To handle them "class HttpHeader"
- GetMimeType
- GetContentType
- GetCookies
 
Commit 11:
------------------------------
**[addons[filesystem] add translator for file read bits**

This also add ADDON_ on begin on addon flags.
There three reasons about them:
1. Prevent conflicts with Kodi's one where defined with "static const unsigned int".
2. To have safe in case Kodi becomes changed and addon overseen.
3. To match the enum begin all there with ADDON_
 
Commit 12:
------------------------------
**[addons][filesystem] add #undef for CreateDirectory and DeleteFile**

There about is on Windows macros where bring in this header conflict.
This undefine it and allow the use within kodi::vfs system.
 
Commit 13:
------------------------------
**[addons][network] remove use of KODI_HANDLE on <kodi/c-api/network.h>**

This allow then also remove use of #include "AddonBase.h" there and
make "C" API checks more easy and prevent not needed compile parts.
 
Commit 14:
------------------------------
**[addons][base] fix possible memory overrun on kodi::Log(...) call**

There was before "vsprintf" used, this now changed to "vsnprintf", where
his available array size is given to him.
 
Commit 15:
------------------------------
**[addons][base] rename ADDON_LOG_... to LOG_... and combine with PVR**

This remove the old and rename the new with remove of ADDON_ at begin.
 
Commit 16:
------------------------------
**[addons][base] add documentation about "kodi::Log(...)" function**

Commit 17:
------------------------------
**[addons][base] move all "C" define's from AddonBase.h to addon_base.h**

This moves it from C++ header to new added right "C" header.
 
Commit 18:
------------------------------
**[addons][base] move ADDON_STATUS from AddonBase.h to addon_base.h**

Also here it is to see as a "C" part, where better on this header.
 
Commit 19:
------------------------------
**[addons] use "clang format off" on dependency check place in versions.h**

The main reason, because cmake uses this area in this form to perform its
addon dependency check.
 
Commit 20:
------------------------------
**[addons][base] don't pass C++ part inside AddonGlobalInterface**

This change the class inserted there to KODI_HANDLE where mean void*.

Before with C++ class inside them it was not usable on "C" alone and
also not match a ABI guideline where only be "C" between two independent
parts.
 
Commit 21:
------------------------------
**[addons][base] separate now all C parts complete to addon_base.h**

Commit 22:
------------------------------
**[addons][base] clang and sort cleanups on AddonBase.h**

This change his code to match complete the clang design.
 
Commit 23:
------------------------------
**[addons] add `const C_STRUCT* GetCStructure() const` and  `operator const C_STRUCT*() const`**

This added to allow in another class where use CStructHdl to get this from his values.
 
Commit 24:
------------------------------
**[addons] add DestroyInstance function to addon base class**

This thought to inform related addon that a instance becomes
destroyed.
 
Commit 25:
------------------------------
**[addons][general] add function to check another addon is available**

This thought for cases as another addon (e.g. PVR) need a special
inputstream addon.

Commit 26:
------------------------------
**[addons] allow ask about Kodi's used addon type version and cleanup Dll**

This add a new interface function where the addon can ask about version
in Kodi. Thought to hold easier for checks in other places of headers.

Further as by this request, an increase of all versions needed, are the exported
"C" functions updated and no more needed removed.

Commit 27:
------------------------------
**[addons] fix open of addon settings dialog via addon**

There was wrongly the "!" missing and always has broken the open.

Commit 28:
------------------------------
**[addons] increase binary API versions**

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
All available addons created and nearly all runtime tested, where was OK on run.

Further one screensaver extended to call mostly all interface functions for tests.
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][filesystem] make interface "C" conform

Before was it not in correct "C" ABI between Kodi and Addon.
This add #ifdef about and change some parts on "C" side to allow
compile there.

AlwinEsch committed:
[addons][filesystem] cleanup namespace kodi::addon use

before was on every place this where makes the lines very long,
this change it to "using namespace kodi::addon" everywhere, where
reduce the size.

AlwinEsch committed:
[addons][filesystem] clang code cleanup

this cleanup the parts to match his style

phunkyfish reviewed (commented):
Looks good, just missing GetCookies() ;)

garbear commented on xbmc/addons/interfaces/Filesystem.cpp:
```diff
@@ -551,32 +748,33 @@ char** Interface_Filesystem::get_property_values(
+      static_cast<CFile*>(file)->GetPropertyValues(internalType, name);
   *numValues = values.size();
-  char **ret = static_cast<char**>(malloc(sizeof(char*)*values.size()));
+  char** ret = static_cast<char**>(malloc(sizeof(char*) * values.size()));
```
Another opportunity for `new`? I saw you made such a change above, just wondering

AlwinEsch replied:
Makes sense to use `new`, place there was come by clang cleanup and overseen it.

Must only learn how is on `**` memory :-).

AlwinEsch replied:
Is then `char** ret = new char*[size];` for there, but can't used as his counterpart use `free()`
Here: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/interfaces/AddonBase.cpp#L455-L465

garbear commented on xbmc/addons/interfaces/Filesystem.h:
```diff
@@ -42,12 +54,12 @@ struct Interface_Filesystem
 
   static bool file_exists(void* kodiBase, const char* filename, bool useCache);
-  static int stat_file(void* kodiBase, const char* filename, struct __stat64* buffer);
+  static bool stat_file(void* kodiBase, const char* filename, struct STAT_STRUCTURE* buffer);
```
Is there another opportunity to change a stat's `int` return type to `bool`? I saw a comparison of a `Stat()` function against `0` above. Just wondering.

AlwinEsch replied:
About them I change it complete, as `void*` is bad for other languages :smile:.
Make it now with them:
```cpp
    bool (*io_control_seek_possible)(void* kodiBase, void* file);
    bool (*io_control_get_cache_status)(void* kodiBase, void* file, struct VFS_IOCTRL_CACHE_STATUS_DATA* status);
    bool (*io_control_set_cache_rate)(void* kodiBase, void* file, unsigned int rate);
    bool (*io_control_set_retry)(void* kodiBase, void* file, bool retry);
```

garbear commented:
No problems jump out at me

AlwinEsch committed:
[addons][filesystem] cleanup documentation

This makes it cleaner and easier on review. Further are the parts
placed in groups.

AlwinEsch committed:
[addons][filesystem] add IoControl functions

Following are added:
- bool IoControlGetSeekPossible()
- bool IoControlGetCacheStatus(CacheStatus& status)
- bool IoControlSetCacheRate(unsigned int rate)
- bool IoControlSetRetry(bool retry)

AlwinEsch committed:
[addons][filesystem] separate "C" and "C++" parts of headers

By them becomes the "C" part placed in own header to see better what language
is where and to make easier for other languages.

AlwinEsch committed:
[addons][filesystem] add some functions to check file path relate to web

Following functions are added:
- bool kodi::vfs::IsInternetStream(const std::string& path, bool strictCheck = false);
- bool kodi::vfs::IsOnLAN(const std::string& path);
- bool kodi::vfs::IsRemote(const std::string& path);
- bool kodi::vfs::IsLocal(const std::string& path);
- bool kodi::vfs::IsURL(const std::string& path);

AlwinEsch committed:
[addons] add "const C_STRUCT* GetCStructure() const" and" operator const C_STRUCT*() const"

This added to allow in another class where use CStructHdl to get this from his values.

AlwinEsch committed:
[addons][filesystem] add more functions to get web content

Following are added:
- GetHttpHeader
  - To handle them "class HttpHeader"
- GetMimeType
- GetContentType
- GetCookies

AlwinEsch committed:
[addons[filesystem] add translator for file read bits

This also add ADDON_ on begin on addon flags.
There three reasons about them:
1. Prevent conflicts with Kodi's one where defined with "static const unsigned int".
2. To have safe in case Kodi becomes changed and addon overseen.
3. To match the enum begin all there with ADDON_

AlwinEsch committed:
[addons][filesystem] add #undef for CreateDirectory and DeleteFile

There about is on Windows macros where bring in this header conflict.
This undefine it and allow the use within kodi::vfs system.

AlwinEsch committed:
[addons][network] remove use of KODI_HANDLE on <kodi/c-api/network.h>

This allow then also remove use of #include "AddonBase.h" there and
make "C" API checks more easy and prevent not needed compile parts.

AlwinEsch committed:
[addons][base] fix possible memory overrun on kodi::Log(...) call

There was before "vsprintf" used, this now changed to "vsnprintf", where
his available array size is given to him.

AlwinEsch committed:
[addons][base] move all "C" define's from AddonBase.h to addon_base.h

This moves it from C++ header to new added right "C" header.

AlwinEsch committed:
[addons][base] move ADDON_STATUS from AddonBase.h to addon_base.h

Also here it is to see as a "C" part, where better on this header.

AlwinEsch committed:
[addons] use "clang format off" on dependency check place in versions.h

The main reason, because cmake uses this area in this form to perform its
addon dependency check.

AlwinEsch committed:
[addons][base] clang and sort cleanups on AddonBase.h

This change his code to match complete the clang design.

AlwinEsch committed:
[addons][base] separate now all C parts complete to addon_base.h

AlwinEsch commented:
Is updated and also now include some base addon parts with cleanups. As with the first parts must every addon updated, makes it sense to include also the other where need the same.

AlwinEsch committed:
[addons] add DestroyInstance function to addon base class

This thought to inform related addon that a instance becomes
destroyed.

AlwinEsch committed:
[addons][base] don't pass C++ part inside AddonGlobalInterface

This change the class inserted there to KODI_HANDLE where mean void*.

Before with C++ class inside them it was not usable on "C" alone and
also not match a ABI guideline where only be "C" between two independent
parts.

AlwinEsch committed:
[addons] increase binary API versions

As this changes related to all places here increase of all.
Thats also why request heavier to make addon updates for all
only one time.

phunkyfish commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/c-api/addon_base.h:
```diff
@@ -0,0 +1,263 @@
+#define ATTRIBUTE_FORCEINLINE inline
+#endif
+
+#ifdef LOG_DEBUG
+#undef LOG_DEBUG
+#endif
+#ifdef LOG_INFO
+#undef LOG_INFO
+#endif
+#ifdef LOG_NOTICE
+#undef LOG_NOTICE
+#endif
+#ifdef LOG_ERROR
+#undef LOG_ERROR
+#endif
```
Is this correct?

AlwinEsch committed:
[addons][base] add documentation about "kodi::Log(...)" function

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch committed:
[addons][base] move ADDON_LOG_... parts to "kodi/c-api/addon_base.h"

As them only "C" it should in "C" parts.

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
Has done the runtime test of mostly all addons (except few PVR) and was OK. Only must be this conflict story between addons fixed, found as newest, if `audiodecoder.ssf` enabled can `vfs.sacd` not play his iso files :unamused:.

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch committed:
[addons] fix open of addon settings dialog via addon

There was wrongly the "!" missing and always has broken the open.

AlwinEsch committed:
[addons][filesystem] add function IsOpen to CFile

This thought to allow other place a check that related file
was really opened.

AlwinEsch committed:
[addons][filesystem] replace __stat64 with STAT_STRUCTURE

This done to have it more OS independent, before was about time in
not complete compatible between.

AlwinEsch commented:
Further tests of interface with addons performed, also mostly all function on AddonBase.h, Filesystem.h and General.h runtime testes.

If jenkins happy is this request ready to go in (after remove of "[TEMP]..." commit).

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
Addon builds was all correct https://jenkins.kodi.tv/job/BuildMultiWithAddons-PR/64/, test patch becomes now removed.

AlwinEsch commented:
jenkins build this please

Rechi commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -422,6 +423,25 @@ int Interface_General::get_global_idle_time(void* kodiBase)
   return g_application.GlobalIdleTime();
 }
 
+bool Interface_General::is_addon_avilable(void* kodiBase, const char* id, char** version, bool* enabled)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -422,6 +423,25 @@ int Interface_General::get_global_idle_time(void* kodiBase)
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || version == nullptr || enabled == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_General::{} - invalid data (addon='{}', version='{}', enabled='{}')", __FUNCTION__,
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_addon_dll.h:
```diff
@@ -14,8 +14,7 @@
 
-  ADDON_STATUS __declspec(dllexport) ADDON_Create(void *callbacks, void* props);
-  ADDON_STATUS __declspec(dllexport) ADDON_CreateEx(void *callbacks, const char* globalApiVersion, void* props);
+  ADDON_STATUS __declspec(dllexport) ADDON_Create(void* callbacks, const char* globalApiVersion, void* props);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

AlwinEsch committed:
[addons] allow ask about Kodi's used addon type version and cleanup Dll

This add a new interface function where the addon can ask about version
in Kodi. Thought to hold easier for checks in other places of headers.

Further as by this request, an increase of all versions needed, are the exported
"C" functions updated and no more needed removed.

AlwinEsch commented:
jenkins build this please

AlwinEsch committed:
[addons][general] add function to check another addon is available

This thought for cases as another addon (e.g. PVR) need a special
inputstream addon.

AlwinEsch commented:
jenkins build this please

phunkyfish commented:
Jenkins build this please

AlwinEsch commented:
Jenkins build this please

AlwinEsch commented:
Jenkins build this please

**There need jenkins a fix** :unamused: :smiley_cat:, a lot of times fails WIN32 or WIN64 with 
- **`Could Not Find f:\builds\workspace\WIN-[32][64]\gtestresults.xml`**

Lets try again :smirk:

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
Jenkins build this please

AlwinEsch commented:
The addon update round can begin.
The jenkins error is unrelated as it is on other Linux parts OK and also the 5 builds was all OK, only bug in Jenkins itself.

basilgello commented:
@ksooo unmerged https://github.com/DigitalDevices/pvr.octonet/pull/42 breaks the pipeline ;)
